### PR TITLE
drop deprecated pointer package in controller

### PR DESCRIPTION
**Pull request by liyuerich**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
use functions in k8s.io/utils/ptr instead: ptr.Deref to dereference a pointer for pkg/controllers & cmd/kube-proxy

#### Which issue(s) this PR fixes:

Ref # https://github.com/kubernetes/utils/pull/283

#### Special notes for your reviewer:
NONE

#### Does this PR introduce a user-facing change?

```
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```
NONE
```

### CODE REVIEW

File: cmd/kube-proxy/app/server.go [hunk 1]
```diff
@@ -179,8 +179,8 @@ func (o *Options) AddFlags(fs *pflag.FlagSet) {
...
 	fs.Int32Var(&o.metricsPort, "metrics-port", o.metricsPort, "The port to bind the metrics server. Use 0 to disable.")
 	fs.MarkDeprecated("metrics-port", "This flag is deprecated and will be removed in a future release. Please use --metrics-bind-address instead.")
-	fs.Int32Var(o.config.OOMScoreAdj, "oom-score-adj", pointer.Int32Deref(o.config.OOMScoreAdj, int32(qos.KubeProxyOOMScoreAdj)), "The oom-score-adj value for kube-proxy process. Values must be within the range [-1000, 1000]. This parameter is ignored if a config file is specified by --config.")
-	fs.Int32Var(o.config.IPTables.MasqueradeBit, "iptables-masquerade-bit", pointer.Int32Deref(o.config.IPTables.MasqueradeBit, 14), "If using the pure iptables proxy, the bit of the fwmark space to mark packets requiring SNAT with.  Must be within the range [0, 31].")
+	fs.Int32Var(o.config.OOMScoreAdj, "oom-score-adj", ptr.Deref[int32](o.config.OOMScoreAdj, int32(qos.KubeProxyOOMScoreAdj)), "The oom-score-adj value for kube-proxy process. Values must be within the range [-1000, 1000]. This parameter is ignored if a config file is specified by --config.")
```
Comment (skitt) [inline cmd/kube-proxy/app/server.go:182]:
```suggestion
	fs.Int32Var(o.config.OOMScoreAdj, "oom-score-adj", ptr.Deref(o.config.OOMScoreAdj, int32(qos.KubeProxyOOMScoreAdj)), "The oom-score-adj value for kube-proxy process. Values must be within the range [-1000, 1000]. This parameter is ignored if a config file is specified by --config.")
```

File: cmd/kube-proxy/app/server.go [hunk 2]
```diff
@@ -179,8 +179,8 @@ func (o *Options) AddFlags(fs *pflag.FlagSet) {
...
+	fs.Int32Var(o.config.IPTables.MasqueradeBit, "iptables-masquerade-bit", ptr.Deref[int32](o.config.IPTables.MasqueradeBit, 14), "If using the pure iptables proxy, the bit of the fwmark space to mark packets requiring SNAT with.  Must be within the range [0, 31].")
```
Comment (skitt) [inline cmd/kube-proxy/app/server.go:183]:
```suggestion
	fs.Int32Var(o.config.IPTables.MasqueradeBit, "iptables-masquerade-bit", ptr.Deref(o.config.IPTables.MasqueradeBit, 14), "If using the pure iptables proxy, the bit of the fwmark space to mark packets requiring SNAT with.  Must be within the range [0, 31].")
```

File: cmd/kube-proxy/app/server.go [hunk 3]
```diff
@@ -203,7 +203,7 @@ func (o *Options) AddFlags(fs *pflag.FlagSet) {
 
 	fs.BoolVar(&o.config.IPVS.StrictARP, "ipvs-strict-arp", o.config.IPVS.StrictARP, "Enable strict ARP by setting arp_ignore to 1 and arp_announce to 2")
 	fs.BoolVar(&o.config.IPTables.MasqueradeAll, "masquerade-all", o.config.IPTables.MasqueradeAll, "If using the pure iptables proxy, SNAT all traffic sent via Service cluster IPs (this not commonly needed)")
-	fs.BoolVar(o.config.IPTables.LocalhostNodePorts, "iptables-localhost-nodeports", pointer.BoolDeref(o.config.IPTables.LocalhostNodePorts, true), "If false Kube-proxy will disable the legacy behavior of allowing NodePort services to be accessed via localhost, This only applies to iptables mode and ipv4.")
+	fs.BoolVar(o.config.IPTables.LocalhostNodePorts, "iptables-localhost-nodeports", ptr.Deref[bool](o.config.IPTables.LocalhostNodePorts, true), "If false Kube-proxy will disable the legacy behavior of allowing NodePort services to be accessed via localhost, This only applies to iptables mode and ipv4.")
```
Comment (skitt) [inline cmd/kube-proxy/app/server.go:206]:
```suggestion
	fs.BoolVar(o.config.IPTables.LocalhostNodePorts, "iptables-localhost-nodeports", ptr.Deref(o.config.IPTables.LocalhostNodePorts, true), "If false Kube-proxy will disable the legacy behavior of allowing NodePort services to be accessed via localhost, This only applies to iptables mode and ipv4.")
```

File: pkg/controller/cronjob/cronjob_controllerv2.go [hunk 4]
```diff
@@ -488,7 +489,7 @@ func (jm *ControllerV2) syncCronJob(
 
 	logger := klog.FromContext(ctx)
 	if cronJob.Spec.TimeZone != nil {
-		timeZone := pointer.StringDeref(cronJob.Spec.TimeZone, "")
+		timeZone := ptr.Deref[string](cronJob.Spec.TimeZone, "")
```
Comment (skitt) [inline pkg/controller/cronjob/cronjob_controllerv2.go:492]:
```suggestion
		timeZone := ptr.Deref(cronJob.Spec.TimeZone, "")
```

File: pkg/controller/cronjob/utils.go [hunk 5]
```diff
@@ -218,7 +218,7 @@ func getJobFromTemplate2(cj *batchv1.CronJob, scheduledTime time.Time) (*batchv1
 
 	if utilfeature.DefaultFeatureGate.Enabled(features.CronJobsScheduledAnnotation) {
 
-		timeZoneLocation, err := time.LoadLocation(pointer.StringDeref(cj.Spec.TimeZone, ""))
+		timeZoneLocation, err := time.LoadLocation(ptr.Deref[string](cj.Spec.TimeZone, ""))
```
Comment (skitt) [inline pkg/controller/cronjob/utils.go:221]:
```suggestion
		timeZoneLocation, err := time.LoadLocation(ptr.Deref(cj.Spec.TimeZone, ""))
```

File: pkg/controller/cronjob/utils_test.go [hunk 6]
```diff
@@ -141,7 +141,7 @@ func TestGetJobFromTemplate2(t *testing.T) {
 			}
 
 			scheduledAnnotation := job.ObjectMeta.Annotations[batchv1.CronJobScheduledTimestampAnnotation]
-			timeZoneLocation, err := time.LoadLocation(pointer.StringDeref(tt.timeZone, ""))
+			timeZoneLocation, err := time.LoadLocation(ptr.Deref[string](tt.timeZone, ""))
```
Comment (skitt) [inline pkg/controller/cronjob/utils_test.go:144]:
```suggestion
			timeZoneLocation, err := time.LoadLocation(ptr.Deref(tt.timeZone, ""))
```

File: pkg/controller/job/backoff_utils.go [hunk 7]
```diff
@@ -207,7 +207,7 @@ func getFinishTimeFromPodReadyFalseCondition(p *v1.Pod) *time.Time {
 
 func getFinishTimeFromDeletionTimestamp(p *v1.Pod) *time.Time {
 	if p.DeletionTimestamp != nil {
-		finishTime := p.DeletionTimestamp.Time.Add(-time.Duration(pointer.Int64Deref(p.DeletionGracePeriodSeconds, 0)) * time.Second)
+		finishTime := p.DeletionTimestamp.Time.Add(-time.Duration(ptr.Deref[int64](p.DeletionGracePeriodSeconds, 0)) * time.Second)
```
Comment (skitt) [inline pkg/controller/job/backoff_utils.go:210]:
```suggestion
		finishTime := p.DeletionTimestamp.Time.Add(-time.Duration(ptr.Deref(p.DeletionGracePeriodSeconds, 0)) * time.Second)
```

File: pkg/controller/job/job_controller.go [hunk 8]
```diff
@@ -1639,7 +1640,7 @@ func (jm *Controller) getPodCreationInfoForIndependentIndexes(logger klog.Logger
 	if len(indexesToAddNow) > 0 {
 		return indexesToAddNow, 0
 	}
-	return indexesToAddNow, pointer.DurationDeref(minRemainingTimePerIndex, 0)
+	return indexesToAddNow, ptr.Deref[time.Duration](minRemainingTimePerIndex, 0)
```
Comment (skitt) [inline pkg/controller/job/job_controller.go:1643]:
```suggestion
	return indexesToAddNow, ptr.Deref(minRemainingTimePerIndex, 0)
```

File: pkg/controller/job/job_controller_test.go [hunk 9]
```diff
@@ -3325,8 +3326,8 @@ func TestSyncJobWithJobPodFailurePolicy(t *testing.T) {
 			if actual.Status.Failed != tc.wantStatusFailed {
 				t.Errorf("unexpected number of failed pods. Expected %d, saw %d\n", tc.wantStatusFailed, actual.Status.Failed)
 			}
-			if pointer.Int32Deref(actual.Status.Terminating, 0) != pointer.Int32Deref(tc.wantStatusTerminating, 0) {
-				t.Errorf("unexpected number of terminating pods. Expected %d, saw %d\n", pointer.Int32Deref(tc.wantStatusTerminating, 0), pointer.Int32Deref(actual.Status.Terminating, 0))
+			if ptr.Deref[int32](actual.Status.Terminating, 0) != ptr.Deref[int32](tc.wantStatusTerminating, 0) {
```
Comment (skitt) [inline pkg/controller/job/job_controller_test.go:3329]:
```suggestion
			if ptr.Deref(actual.Status.Terminating, 0) != ptr.Deref(tc.wantStatusTerminating, 0) {
```

File: pkg/controller/job/job_controller_test.go [hunk 10]
```diff
@@ -3325,8 +3326,8 @@ func TestSyncJobWithJobPodFailurePolicy(t *testing.T) {
...
+				t.Errorf("unexpected number of terminating pods. Expected %d, saw %d\n", ptr.Deref[int32](tc.wantStatusTerminating, 0), ptr.Deref[int32](actual.Status.Terminating, 0))
```
Comment (skitt) [inline pkg/controller/job/job_controller_test.go:3330]:
```suggestion
				t.Errorf("unexpected number of terminating pods. Expected %d, saw %d\n", ptr.Deref(tc.wantStatusTerminating, 0), ptr.Deref(actual.Status.Terminating, 0))
```

File: pkg/controller/resourceclaim/controller.go [hunk 11]
```diff
@@ -904,7 +905,7 @@ func (ec *Controller) syncClaim(ctx context.Context, namespace, name string) err
 
 func owningPod(claim *resourcev1alpha2.ResourceClaim) (string, types.UID) {
 	for _, owner := range claim.OwnerReferences {
-		if pointer.BoolDeref(owner.Controller, false) &&
+		if ptr.Deref[bool](owner.Controller, false) &&
```
Comment (skitt) [inline pkg/controller/resourceclaim/controller.go:908]:
```suggestion
		if ptr.Deref(owner.Controller, false) &&
```

**Review by skitt — COMMENTED** [review]
As before, please don’t specify the type when it can be inferred.

File: cmd/kube-proxy/app/server.go (hunk 1 above)
Comment (kannon92) [reply cmd/kube-proxy/app/server.go:182]: Looking at the type for OOMScoreAdj, I don't think you need the type case.

```suggestion
	fs.Int32Var(o.config.OOMScoreAdj, "oom-score-adj", ptr.Deref(o.config.OOMScoreAdj, qos.KubeProxyOOMScoreAdj), "The oom-score-adj value for kube-proxy process. Values must be within the range [-1000, 1000]. This parameter is ignored if a config file is specified by --config.")
```

**Comment by bart0sh** [comment]
/triage accepted
/priority important-longterm

@liyuerich please, fix CI failures, thanks.

**Comment by iiiceoo** [comment]
@liyuerich Could you please modify the test files under cmd/kube-proxy as well:

https://github.com/kubernetes/kubernetes/blob/0241da314e0e69817d66313b45a69c19d1ce7327/cmd/kube-proxy/app/server_test.go#L39

https://github.com/kubernetes/kubernetes/blob/0241da314e0e69817d66313b45a69c19d1ce7327/cmd/kube-proxy/app/server_others_test.go#L43

**Comment by liyuerich** [comment]
@skitt 
sorry, it seems the split operation revert all my changes "type when it can be inferred." I will update again

**Comment by liyuerich** [comment]
> @liyuerich Could you please modify the test files under cmd/kube-proxy as well:
> 
> https://github.com/kubernetes/kubernetes/blob/0241da314e0e69817d66313b45a69c19d1ce7327/cmd/kube-proxy/app/server_test.go#L39
> 
> https://github.com/kubernetes/kubernetes/blob/0241da314e0e69817d66313b45a69c19d1ce7327/cmd/kube-proxy/app/server_others_test.go#L43

@iiiceoo 
May I leave this PR to handle it?

https://github.com/kubernetes/kubernetes/pull/120565

File: pkg/controller/job/job_controller.go [hunk 12]
```diff
@@ -56,6 +56,12 @@ import (
 	"k8s.io/utils/ptr"
 )
 
+const (
+	// PodFailurePolicy reason indicates a job failure condition is added due to
+	// a failed pod matching a pod failure policy rule
+	jobConditionReasonPodFailurePolicy = "PodFailurePolicy"
+)
+
```
Comment (skitt) [inline pkg/controller/job/job_controller.go:64]: This was removed, please don’t add it back.

**Review by skitt — COMMENTED** [review]
Please squash your commits into one (rebase and mark your last commit for squashing; that will also get rid of the merge commit).

File: pkg/controller/job/job_controller.go (hunk 12 above)
Comment (liyuerich) [reply pkg/controller/job/job_controller.go]: yes, remove is done.

**Comment by liyuerich** [comment]
> Please squash your commits into one (rebase and mark your last commit for squashing; that will also get rid of the merge commit).

yes, it is done

File: pkg/controller/cronjob/cronjob_controllerv2.go [hunk 13]
```diff
@@ -48,6 +48,7 @@ import (
 	"k8s.io/kubernetes/pkg/controller"
 	"k8s.io/kubernetes/pkg/controller/cronjob/metrics"
 	"k8s.io/utils/pointer"
```
Comment (bowei) [inline pkg/controller/cronjob/cronjob_controllerv2.go:50]: I see both k8s.io/utils/pointer and k8s.io/utils/ptr.

Are both needed?

Comment (bowei) [reply pkg/controller/cronjob/cronjob_controllerv2.go:50]: From what I can tell, the `ptr` package should replace the `pointer` package.

Comment (liyuerich) [reply pkg/controller/cronjob/cronjob_controllerv2.go:50]: bowei,  both are needed. you could find "pointer.StringEqual" still used in this file and there is another PR is working on this:

https://github.com/kubernetes/kubernetes/pull/120629

**Comment by liyuerich** [comment]
@bart0sh please review . any comment and feedback, please let me know. thanks.

**Comment by liyuerich** [comment]
@bart0sh , could you help to review ? thanks.

**Comment by danwinship** [comment]
/remove-sig network
/remove-area kube-proxy

(no longer contains any kube-proxy changes)

**Comment by bart0sh** [comment]
/assign @deads2k @mikedanese @andrewsykim
for approval

**Comment by liyuerich** [comment]
> /assign @deads2k @mikedanese @andrewsykim for approval

@deads2k @mikedanese @andrewsykim could you help to approve? thanks.

File: pkg/controller/resourceclaim/controller.go [hunk 14]
```diff
@@ -46,6 +46,7 @@ import (
 	podutil "k8s.io/kubernetes/pkg/api/v1/pod"
 	"k8s.io/kubernetes/pkg/controller/resourceclaim/metrics"
 	"k8s.io/utils/pointer"
```
Comment (kannon92) [inline pkg/controller/resourceclaim/controller.go:48]: Could you just remove this dependency?

File: pkg/controller/cronjob/cronjob_controllerv2.go (hunk 13 above)
Comment (kannon92) [inline pkg/controller/cronjob/cronjob_controllerv2.go:50]: Same here. Let's drop pointer on the files that you are updating.

File: pkg/controller/resourceclaim/controller.go (hunk 14 above)
Comment (liyuerich) [reply pkg/controller/resourceclaim/controller.go]: thanks reply.  it is done.

File: pkg/controller/cronjob/cronjob_controllerv2.go (hunk 13 above)
Comment (liyuerich) [reply pkg/controller/cronjob/cronjob_controllerv2.go]: thanks reply.  it is done.

**Comment by liyuerich** [comment]
@bart0sh  @bowei please help to review . thanks.

**Comment by liyuerich** [comment]
@bart0sh  @pohly please help to review and approve. thanks.

**Review by pohly — APPROVED** [review]
/lgtm

**Comment by pohly** [comment]
/assign @janetkuo

For cronjob controller approval.

**Comment by liyuerich** [comment]
thanks @carlory and @pohly for your comments.

@janetkuo , could you help to approve? thanks.

**Comment by liyuerich** [comment]
@janetkuo  
Janet， could you help to approve? thanks.

**Review by janetkuo — APPROVED** [review]

**Pull request merged**